sparse/_coo/numba_extension.py

```diff
@@ -0,0 +1,165 @@
+"""
+Numba support for COO objects.
+
+For now, this just supports attribute access
+"""
+import numpy as np
+import numba
+from numba.extending import (
+    models,
+    register_model,
+    box,
+    unbox,
+    NativeValue,
+    make_attribute_wrapper,
+    type_callable,
+    lower_builtin,
+)
+from numba.targets.imputils import impl_ret_borrowed, lower_constant
+from numba.typing.typeof import typeof_impl
+from numba import cgutils, types
+from sparse._utils import _zero_of_dtype
+
+from . import COO
+
+__all__ = ["COOType"]
+
+
+class COOType(types.Type):
+    def __init__(self, data_dtype: np.dtype, coords_dtype: np.dtype, ndim: int):
+        assert isinstance(data_dtype, np.dtype)
+        assert isinstance(coords_dtype, np.dtype)
+        self.data_dtype = data_dtype
+        self.coords_dtype = coords_dtype
+        self.ndim = ndim
+        super().__init__(
+            name="COOType[{!r}, {!r}, {!r}]".format(
+                numba.from_dtype(data_dtype), numba.from_dtype(coords_dtype), ndim
+            )
+        )
+
+    @property
+    def key(self):
+        return self.data_dtype, self.coords_dtype, self.ndim
+
+    @property
+    def data_type(self):
+        return numba.from_dtype(self.data_dtype)[:]
+
+    @property
+    def coords_type(self):
+        return numba.from_dtype(self.coords_dtype)[:, :]
+
+    @property
+    def shape_type(self):
+        return types.UniTuple(types.int64, self.ndim)
+
+    @property
+    def fill_value_type(self):
+        return numba.from_dtype(self.data_dtype)
+
+
+@typeof_impl.register(COO)
+def _typeof_COO(val: COO, c) -> COOType:
+    return COOType(
+        data_dtype=val.data.dtype, coords_dtype=val.coords.dtype, ndim=val.ndim
+    )
+
+
+@register_model(COOType)
+class COOModel(models.StructModel):
+    def __init__(self, dmm, fe_type):
+        members = [
+            ("data", fe_type.data_type),
+            ("coords", fe_type.coords_type),
+            ("shape", fe_type.shape_type),
+            ("fill_value", fe_type.fill_value_type),
+        ]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+@type_callable(COO)
+def type_COO(context):
+    # TODO: accept a fill_value kwarg
+    def typer(coords, data, shape):
+        return COOType(
+            coords_dtype=numba.numpy_support.as_dtype(coords.dtype),
+            data_dtype=numba.numpy_support.as_dtype(data.dtype),
+            ndim=len(shape),
+        )
+
+    return typer
+
+
+@lower_builtin(COO, types.Any, types.Any, types.Any)
+def impl_COO(context, builder, sig, args):
+    typ = sig.return_type
+    coords, data, shape = args
+    coo = cgutils.create_struct_proxy(typ)(context, builder)
+    coo.coords = coords
+    coo.data = data
+    coo.shape = shape
+    coo.fill_value = context.get_constant_generic(
+        builder, typ.fill_value_type, _zero_of_dtype(typ.data_dtype)
+    )
+    return impl_ret_borrowed(context, builder, sig.return_type, coo._getvalue())
```

Comment:
@sklam suggested that this might be possible to do using `@overload`, which would make the `fill_value` kwargs easy.

Would you be able to give an example?

Reply:
Maybe by `@overload`-ing the `COO` type itself?

Reply:
Right, but in an overload context I don't have access to `builder` and`context`, so can't call `cgutils.create_struct_proxy(typ)`

Reply:
@eric-wieser not sure what the conversation with @sklam was, however, [`@intrinsic`](http://numba.pydata.org/numba-doc/latest/extending/high-level.html#implementing-intrinsics) is what you need if you want to use `@overload` and also have the `context`/`builder` accessible.

Reply:
I attempted this as:
```
@intrinsic
def _make_coo(typingctx, coords, data, shape, fill_value):
    # TODO: accept a fill_value kwarg
    result_type = COOType(
        coords_dtype=numba.numpy_support.as_dtype(coords.dtype),
        data_dtype=numba.numpy_support.as_dtype(data.dtype),
        ndim=len(shape),
    )
    args_types = [
        result_type.coords_dtype,
        result_type.data_dtype,
        result_type.shape_type,
        result_type.fill_value_type
    ]
    if fill_value == types.none:
        args_types[-1] = types.none
        def codegen(context, builder, sig, args):
            typ = sig.return_type
            coords, data, shape, fill_value = args
            coo = cgutils.create_struct_proxy(typ)(context, builder)
            coo.coords = coords
            coo.data = data
            coo.shape = shape
            coo.fill_value = context.get_constant_generic(
                builder, typ.fill_value_type, _zero_of_dtype(typ.fill_value_type)
            )
            return impl_ret_borrowed(context, builder, sig.return_type, coo._getvalue())
    else:
        def codegen(context, builder, sig, args):
            typ = sig.return_type
            coords, data, shape, fill_value = args
            coo = cgutils.create_struct_proxy(typ)(context, builder)
            coo.coords = coords
            coo.data = data
            coo.shape = shape
            coo.fill_value = fill_value
            return impl_ret_borrowed(context, builder, sig.return_type, coo._getvalue())

    return sig, codegen


@overload(COO)
def impl_COO(coords, data, shape, *, fill_value=None):
    # convert a keyword to positional argument
    return _make_coo(coords, data, shape, fill_value)
```
but got the error
```
Invalid use of Function(<class 'sparse._coo.core.COO'>) with argument(s) of type(s): (array(int64, 2d, A), array(int32, 1d, A), UniTuple(int64 x 1))
 * parameterized
In definition 0:
    NotImplementedError: <intrinsic _make_coo> is not usable in pure-python
    raised from C:\Users\wiese\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\extending.py:304
In definition 1:
    NotImplementedError: <intrinsic _make_coo> is not usable in pure-python
    raised from C:\Users\wiese\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\extending.py:304
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<class 'sparse._coo.core.COO'>)
[2] During: typing of call at <ipython-input-2-523b502ede43> (5)


File "<ipython-input-2-523b502ede43>", line 5:
def get_data(c):
    return sparse.COO(c.coords, c.data, c.shape)
```

Let's leave this to a follow-up, there's enough complexity here already.

Reply:
Resolving, I don't think keyword-only args are possible at all in the latest numba.

Reply:
There's a problem in here: https://github.com/numba/numba/blob/master/numba/typing/templates.py#L154 that will prevent it from working. Resolving as-is is probably the best option for now.